In [41]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
pd.read_csv("general_data.csv")

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [45]:
dataset = pd.read_csv("general_data.csv")

In [46]:
dataset.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [48]:
dataset.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [50]:
dataset.groupby(by="Attrition").size()

Attrition
No     3699
Yes     711
dtype: int64

In [52]:
dataset.loc[dataset['Attrition'] == 'Yes']

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                   10          1  Life Sciences              1           2   
6                   11          2        Medical              1           7   
13                   1          1        Medical              1          14   
28                   1          2        Medical              1          29   
30                   4          3        Medical              1          31   
...                ...        ...            ...            ...         ...   
4381                 7          1  Life Sciences              1        4382   
4386                11          4      Marketing              1        4387   
4388                 1          3  Life Sciences              1        4389   
4391                23          1  Life Sciences              1        4392   
4402                 2          3      Marketing              1        4403   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1     Female  ...                 0.0      Y                23              8   
6       Male  ...                 2.0      Y                20              8   
13      Male  ...                 1.0      Y                11              8   
28      Male  ...                 3.0      Y                14              8   
30      Male  ...                 2.0      Y                11              8   
...      ...  ...                 ...    ...               ...            ...   
4381  Female  ...                 1.0      Y                21              8   
4386    Male  ...                 5.0      Y                21              8   
4388    Male  ...                 7.0      Y                11              8   
4391    Male  ...                 0.0      Y                11              8   
4402    Male  ...                 6.0      Y                11              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
1                    1               6.0                      3   
6                    1               5.0                      2   
13                   2              10.0                      4   
28                   1              19.0                      2   
30                   0               5.0                      5   
...                ...               ...                    ...   
4381                 0               4.0                      2   
4386                 0               8.0                      2   
4388                 0              13.0                      2   
4391                 0               4.0                      2   
4402                 1              17.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1                  5                        1                     4  
6                  0                        0                     0  
13                10                        9                     9  
28                 1                        0                     0  
30                 3                        0                     2  
...              ...                      ...                   ..

In [53]:
dataset1 = dataset.loc[dataset['Attrition'] == 'Yes']

In [54]:
dataset1.head()

Age Attrition     BusinessTravel              Department  \
1    31       Yes  Travel_Frequently  Research & Development   
6    28       Yes      Travel_Rarely  Research & Development   
13   47       Yes         Non-Travel  Research & Development   
28   44       Yes  Travel_Frequently  Research & Development   
30   26       Yes      Travel_Rarely  Research & Development   

    DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                 10          1  Life Sciences              1           2   
6                 11          2        Medical              1           7   
13                 1          1        Medical              1          14   
28                 1          2        Medical              1          29   
30                 4          3        Medical              1          31   

    Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1   Female  ...                 0.0      Y                23              8   
6     Male  ...                 2.0      Y                20              8   
13    Male  ...                 1.0      Y                11              8   
28    Male  ...                 3.0      Y                14              8   
30    Male  ...                 2.0      Y                11              8   

    StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
1                  1               6.0                      3               5   
6                  1               5.0                      2               0   
13                 2              10.0                      4              10   
28                 1              19.0                      2               1   
30                 0               5.0                      5               3   

    YearsSinceLastPromotion  YearsWithCurrManager  
1                         1                     4  
6                         0                     0  
13                        9                     9  
28                        0                     0  
30                        0                     2  

[5 rows x 24 columns]

In [55]:
dataset1[['Age','DistanceFromHome','Education','PercentSalaryHike','StandardHours']].describe()

Age  DistanceFromHome   Education  PercentSalaryHike  \
count  711.000000        711.000000  711.000000         711.000000   
mean    33.607595          9.012658    2.877637          15.481013   
std      9.675693          7.772368    1.014233           3.775289   
min     18.000000          1.000000    1.000000          11.000000   
25%     28.000000          2.000000    2.000000          12.000000   
50%     32.000000          7.000000    3.000000          14.000000   
75%     39.000000         15.000000    4.000000          18.000000   
max     58.000000         29.000000    5.000000          25.000000   

       StandardHours  
count          711.0  
mean             8.0  
std              0.0  
min              8.0  
25%              8.0  
50%              8.0  
75%              8.0  
max              8.0

In [57]:
dataset1[['TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager']].describe()

TotalWorkingYears  YearsAtCompany  YearsSinceLastPromotion  \
count         709.000000      711.000000               711.000000   
mean            8.255289        5.130802                 1.945148   
std             7.164018        5.941598                 3.148633   
min             0.000000        0.000000                 0.000000   
25%             3.000000        1.000000                 0.000000   
50%             7.000000        3.000000                 1.000000   
75%            10.000000        7.000000                 2.000000   
max            40.000000       40.000000                15.000000   

       YearsWithCurrManager  
count            711.000000  
mean               2.852321  
std                3.138918  
min                0.000000  
25%                0.000000  
50%                2.000000  
75%                5.000000  
max               14.000000

In [58]:
dataset1[['Age','DistanceFromHome','Education','PercentSalaryHike','StandardHours']].mode()

Age  DistanceFromHome  Education  PercentSalaryHike  StandardHours
0   29               2.0        3.0               13.0            8.0
1   31               NaN        NaN                NaN            NaN

In [59]:
dataset1[['TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager']].mode()

TotalWorkingYears  YearsAtCompany  YearsSinceLastPromotion  \
0                1.0               1                        0   

   YearsWithCurrManager  
0                     0

In [60]:
dataset1[['Age','DistanceFromHome','Education','PercentSalaryHike','StandardHours']].skew()

Age                  0.712699
DistanceFromHome     0.964420
Education           -0.216055
PercentSalaryHike    0.763693
StandardHours        0.000000
dtype: float64

In [61]:
dataset1[['TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager']].skew()

TotalWorkingYears          1.679324
YearsAtCompany             2.670876
YearsSinceLastPromotion    2.208164
YearsWithCurrManager       1.025486
dtype: float64

In [62]:
dataset1[['Age','DistanceFromHome','Education','PercentSalaryHike','StandardHours']].kurt()

Age                 -0.073144
DistanceFromHome    -0.041771
Education           -0.616494
PercentSalaryHike   -0.418411
StandardHours        0.000000
dtype: float64

In [63]:
dataset1[['TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager']].kurt()

TotalWorkingYears          3.703175
YearsAtCompany             9.456107
YearsSinceLastPromotion    4.775929
YearsWithCurrManager       0.242923
dtype: float64

## Hypothesis Statements

__Employees with the mean education of 2.8 years and standard deviation of 1.01 has the highest number of attrition__

__Employees with the mean age of 33.6 and standard deviation of 9.67 has the highest number of attrition__

__Employees with the mean Distance from home as 9.01 and standard deviation of 7 has the highest number of attrition__

__Employees with the mean Percent Salary Hike as 15.48 and standard deviation of 3.77 has the highest number of attrition__

__Employees who are 29 years old has higher number of attrition compared to others__ 

__Employees who has 1 year of experience with the compary has higher number of attritions__

__Employees with the mean Years at company as 5.13 and standard deviation of 5.94 has the highest number of attrition__

__Employees who get salary hike of 13% has highest number of attrition__

__Employees who live 2 kms away from the office has the highest number of attrition__

__Employees with the total working experience of 1 year has the higher number of attrition__